source/ShipJumpNavigation.cpp

```diff
@@ -0,0 +1,223 @@
+
+	// Make it possible for a hyperdrive or jump drive to be integrated into a ship.
+	ParseOutfit(ship->BaseAttributes(), shipHyperCost);
+	// Check each outfit from this ship to determine if it has jump capabilities.
+	for(const auto &it : ship->Outfits())
+		ParseOutfit(*it.first, shipHyperCost);
+}
+
+
+
+// Get the amount of fuel that would be expended to jump to the destination. If the destination is
+// nullptr then return the maximum amount of fuel that this ship could expend in one jump.
+double ShipJumpNavigation::JumpFuel(const System *destination) const
+{
+	assert(ship && "A jump navigation's ship cannot be null.");
+	const System *currentSystem = ship->GetSystem();
+	// A currently-carried ship requires no fuel to jump, because it cannot jump.
+	if(!currentSystem)
+		return 0.;
+
+	// If no destination is given, return the maximum fuel per jump.
+	if(!destination)
+		return max(JumpDriveFuel(maxJumpRange), HyperdriveFuel());
+
+	return GetCheapestJumpType(destination).second;
+}
+
+
+
+// Get the maximum distance that this ship can jump.
+double ShipJumpNavigation::JumpRange() const
+{
+	return maxJumpRange;
+}
+
+
+
+// Get the cost of making a jump of the given type (if possible). Returns 0 if the jump can't be made.
+double ShipJumpNavigation::HyperdriveFuel() const
+{
+	// If this ship doesn't have a hyperdrive then hyperdriveCost will already be 0.
+	return hyperdriveCost;
+}
+
+
+
+double ShipJumpNavigation::JumpDriveFuel(double distance) const
+{
+	// If this ship has no jump drive then return 0.
+	if(!hasJumpDrive)
+		return 0.;
+	// If this exact distance is in the list then return its cost. This will
+	// likely only occur if the given distance is 0 or maxJumpRange.
+	if(jumpDriveCosts.count(distance))
+		return jumpDriveCosts.find(distance)->second;
+	// Otherwise, find the first jump range that covers the distance. Iterate over
+	// the costs map until we reach the jump range just above the given distance.
+	auto it = std::find_if(jumpDriveCosts.begin(), jumpDriveCosts.end(),
+		[distance](const pair<double, double> &range) -> bool { return range.first > distance; });
+	return (it == jumpDriveCosts.end()) ? 0. : it->second;
+}
+
+
+
+// Get the cheapest jump method and its cost for a jump to the destination system.
+// If no jump method is possible, returns JumpType::None with a jump cost of 0.
+pair<JumpType, double> ShipJumpNavigation::GetCheapestJumpType(const System *destination) const
+{
+	assert(ship && "A jump navigation's ship cannot be null.");
+	return GetCheapestJumpType(ship->GetSystem(), destination);
+}
+
+
+
+pair<JumpType, double> ShipJumpNavigation::GetCheapestJumpType(const System *from, const System *to) const
+{
+	bool linked = from->Links().count(to);
+	double hyperFuelNeeded = HyperdriveFuel();
+	double jumpFuelNeeded = JumpDriveFuel((linked || from->JumpRange())
+			? 0. : from->Position().Distance(to->Position()));
```

Comment:
Can you explain what's going on in these two lines? It doesn't seem right to me.

Reply:
This is just code that's been moved around since #5193. What's happening here is that we're getting the jump fuel needed for the jump range between the systems `from` and `to`, but if the systems are linked or if the system `from` has its own jump range, we use the cheapest jump drive possible, which gets mapped to distance 0.

The lore reason that the cheapest JD is used for linked systems is because the link acts to reduce the resistance between systems `from` and `to`, allowing you to use the cheapest JD instead of needing to use whatever JD matches the distance between the systems, which may be more expensive.
The gameplay reason that we pass 0 for linked systems is so that two systems which are linked but have a distance farther than 100 don't prevent you from jumping.
There's no lore reason for why systems with their own jump range map to a distance of 0 to use the cheapest jump drive, I just did that for ease of implementation.

Reply:
Ok, the thing with the jump range was what I was wondering about.
A comment explaining that might be useful because I don't think it's entirely clear that that's how a system with it's own jump range is handled.